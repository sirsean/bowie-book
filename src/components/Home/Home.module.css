.container {
  width: 100%;
  min-height: 100%;
  padding: 1rem;
  text-align: center;
  background: var(--color-background-gradient);
  background-size: 400% 400%;
  animation: gradientBG 15s ease infinite;
}

@keyframes gradientBG {
  0% { background-position: 0% 50% }
  50% { background-position: 100% 50% }
  100% { background-position: 0% 50% }
}

.title {
  font-size: 3rem;
  margin: 1.5rem 0 2.5rem;
  background: var(--rainbow-gradient);
  background-clip: text;
  -webkit-background-clip: text;
  color: transparent;
  text-shadow: 3px 3px 6px rgba(0, 0, 0, 0.3);
  position: relative;
  display: inline-block;
  font-weight: bold;
  padding: 0.5rem 1rem;
}

.title::after {
  content: '';
  position: absolute;
  bottom: 0;
  left: 25%;
  width: 50%;
  height: 4px;
  background: var(--rainbow-gradient);
  border-radius: 2px;
}

.imageGrid {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
  gap: 1.5rem;
  width: 100%;
  max-width: 1200px;
  margin: 0 auto;
  padding: 1rem;
}

.gridItem {
  position: relative;
  border-radius: var(--border-radius-lg);
  overflow: hidden;
  box-shadow: var(--rainbow-shadow);
  aspect-ratio: 1 / 1;
  transition: transform 0.3s ease, box-shadow 0.3s ease;
  border: 4px solid white;
}

.gridItem:hover, .gridItem:focus-within {
  transform: translateY(-10px) scale(1.03);
  box-shadow: 0 15px 30px rgba(155, 93, 229, 0.6);
}

.gridItem:nth-child(1) { border-color: var(--color-red); }
.gridItem:nth-child(2) { border-color: var(--color-orange); }
.gridItem:nth-child(3) { border-color: var(--color-green); }
.gridItem:nth-child(4) { border-color: var(--color-purple); }

.gridItem a {
  display: block;
  width: 100%;
  height: 100%;
  position: relative;
  overflow: hidden;
}

.gridItem img {
  width: 100%;
  height: 100%;
  object-fit: cover;
  transition: transform 0.5s ease;
}

.gridItem:hover img {
  transform: scale(1.1);
}

.gridItemTitle {
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  background: linear-gradient(to top, rgba(0,0,0,0.8), transparent);
  color: white;
  padding: 1.5rem 0.5rem 0.5rem;
  text-align: center;
  font-size: 1.2rem;
  font-weight: bold;
  text-shadow: var(--text-shadow);
  transition: opacity 0.3s ease;
}

/* Tablet (iPad) */
@media (max-width: 768px) {
  .imageGrid {
    grid-template-columns: repeat(2, 1fr);
  }
}

/* Mobile */
@media (max-width: 480px) {
  .imageGrid {
    grid-template-columns: 1fr;
    gap: 1.5rem;
  }
  
  .title {
    font-size: 2rem;
    margin: 0.5rem 0 1.5rem;
  }
}