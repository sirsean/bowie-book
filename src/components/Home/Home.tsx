import { Link } from 'react-router-dom';
import styles from './Home.module.css';

interface BookPreview {
  id: string;
  title: string;
  coverImage: string;
}

const books: BookPreview[] = [
  {
    id: 'bonne-adventure',
    title: 'Bonne Adventure',
    coverImage: '/books/bonne-adventure/0-cover.webp'
  },
  {
    id: 'dragon-fighter',
    title: 'Dragon Fighter',
    coverImage: '/books/dragon-fighter/0-cover.webp'
  },
  {
    id: 'skyward-bound',
    title: 'Skyward Bound',
    coverImage: '/books/skyward-bound/0.webp'
  },
  {
    id: 'ziggy-the-bunny',
    title: 'Ziggy the Bunny',
    coverImage: '/books/ziggy-the-bunny/0-cover.webp'
  }
];

/**
 * Home component displays a grid of book covers with links to each book
 */
const Home = (): JSX.Element => {
  return (
    <div className={styles.container}>
      <h1 className={styles.title}>Bowie's Books!</h1>
      <div className={styles.imageGrid}>
        {books.map((book) => (
          <div key={book.id} className={styles.gridItem}>
            <Link to={`/${book.id}`}>
              <img 
                src={book.coverImage} 
                alt={`${book.title} Cover`} 
                loading="lazy"
              />
              <div className={styles.gridItemTitle}>{book.title}</div>
            </Link>
          </div>
        ))}
      </div>
    </div>
  );
};

export default Home;