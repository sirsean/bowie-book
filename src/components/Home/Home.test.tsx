import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import Home from './Home';

const renderWithRouter = (initialEntries: string[] = ['/']) => {
  return render(
    <MemoryRouter initialEntries={initialEntries}>
      <Home />
    </MemoryRouter>
  );
};

describe('Home Component', () => {
  describe('Book Cards Rendering', () => {
    it('renders 6 book cards with correct cover images and alt text', () => {
      renderWithRouter();

      // Check that 6 book covers are rendered
      const bookCovers = screen.getAllByRole('img');
      expect(bookCovers).toHaveLength(6);

      // Verify each book cover has correct src and alt text
      const expectedBooks = [
        { title: 'Bonne Adventure', coverImage: '/books/bonne-adventure/0-cover.webp' },
        { title: 'Dragon Fighter', coverImage: '/books/dragon-fighter/0-cover.webp' },
        { title: 'Skyward Bound', coverImage: '/books/skyward-bound/0.webp' },
        { title: 'Ziggy the Bunny', coverImage: '/books/ziggy-the-bunny/0-cover.webp' },
        { title: 'Super Bowie', coverImage: '/books/super-bowie/0-cover.jpg' },
        {
          title: 'Superkitty Saves Bunnytown',
          coverImage: '/books/superkitty-saves-bunnytown/superkitty-cover.png',
        },
      ];

      expectedBooks.forEach((book, index) => {
        const bookCover = bookCovers[index];
        expect(bookCover).toHaveAttribute('src', book.coverImage);
        expect(bookCover).toHaveAttribute('alt', `${book.title} Cover`);
      });
    });

    it('renders book titles under each image', () => {
      renderWithRouter();

      // Check that all book titles are rendered
      const expectedTitles = [
        'Bonne Adventure',
        'Dragon Fighter',
        'Skyward Bound',
        'Ziggy the Bunny',
        'Super Bowie',
        'Superkitty Saves Bunnytown',
      ];

      expectedTitles.forEach((title) => {
        expect(screen.getByText(title)).toBeInTheDocument();
      });
    });

    it('renders main title "Bowie\'s Books!"', () => {
      renderWithRouter();

      expect(screen.getByText("Bowie's Books!")).toBeInTheDocument();
    });
  });

  describe('Navigation', () => {
    it('clicking on a book cover navigates to the correct book page', async () => {
      const user = userEvent.setup();
      renderWithRouter();

      // Test clicking on the first book (Bonne Adventure)
      const bonneAdventureLink = screen.getByRole('link', {
        name: /bonne adventure cover bonne adventure/i,
      });
      await user.click(bonneAdventureLink);

      // Check that the link has the correct href
      expect(bonneAdventureLink).toHaveAttribute('href', '/bonne-adventure');
    });

    it('clicking on Dragon Fighter cover navigates to dragon-fighter page', async () => {
      const user = userEvent.setup();
      renderWithRouter();

      const dragonFighterLink = screen.getByRole('link', {
        name: /dragon fighter cover dragon fighter/i,
      });
      await user.click(dragonFighterLink);

      expect(dragonFighterLink).toHaveAttribute('href', '/dragon-fighter');
    });

    it('clicking on Skyward Bound cover navigates to skyward-bound page', async () => {
      const user = userEvent.setup();
      renderWithRouter();

      const skywardBoundLink = screen.getByRole('link', {
        name: /skyward bound cover skyward bound/i,
      });
      await user.click(skywardBoundLink);

      expect(skywardBoundLink).toHaveAttribute('href', '/skyward-bound');
    });

    it('clicking on Ziggy the Bunny cover navigates to ziggy-the-bunny page', async () => {
      const user = userEvent.setup();
      renderWithRouter();

      const ziggyLink = screen.getByRole('link', {
        name: /ziggy the bunny cover ziggy the bunny/i,
      });
      await user.click(ziggyLink);

      expect(ziggyLink).toHaveAttribute('href', '/ziggy-the-bunny');
    });

    it('clicking on Super Bowie cover navigates to super-bowie page', async () => {
      const user = userEvent.setup();
      renderWithRouter();

      const superBowieLink = screen.getByRole('link', { name: /super bowie cover super bowie/i });
      await user.click(superBowieLink);

      expect(superBowieLink).toHaveAttribute('href', '/super-bowie');
    });

    it('clicking on Superkitty Saves Bunnytown cover navigates to superkitty-saves-bunnytown page', async () => {
      const user = userEvent.setup();
      renderWithRouter();

      const superkittyLink = screen.getByRole('link', {
        name: /superkitty saves bunnytown cover superkitty saves bunnytown/i,
      });
      await user.click(superkittyLink);

      expect(superkittyLink).toHaveAttribute('href', '/superkitty-saves-bunnytown');
    });
  });

  describe('Accessibility', () => {
    it('all book covers have proper alt text for screen readers', () => {
      renderWithRouter();

      const expectedAltTexts = [
        'Bonne Adventure Cover',
        'Dragon Fighter Cover',
        'Skyward Bound Cover',
        'Ziggy the Bunny Cover',
        'Super Bowie Cover',
        'Superkitty Saves Bunnytown Cover',
      ];

      expectedAltTexts.forEach((altText) => {
        expect(screen.getByAltText(altText)).toBeInTheDocument();
      });
    });

    it('all book links are keyboard accessible', async () => {
      renderWithRouter();

      const links = screen.getAllByRole('link');
      expect(links).toHaveLength(6);

      links.forEach((link) => {
        expect(link).toHaveAttribute('href');
        expect(link.getAttribute('href')).toMatch(/^\/[a-z-]+$/);
      });
    });
  });

  describe('Image Loading', () => {
    it('sets lazy loading on all book cover images', () => {
      renderWithRouter();

      const bookCovers = screen.getAllByRole('img');
      bookCovers.forEach((cover) => {
        expect(cover).toHaveAttribute('loading', 'lazy');
      });
    });
  });

  describe('Grid Layout', () => {
    it('renders books in a grid container', () => {
      const { container } = renderWithRouter();

      // Check that the grid structure is present
      const gridContainer = container.querySelector('[class*="container"]');
      expect(gridContainer).toBeInTheDocument();

      // Check that the image grid is present
      const imageGrid = container.querySelector('[class*="imageGrid"]');
      expect(imageGrid).toBeInTheDocument();

      // Check that grid items are present (only direct children of imageGrid)
      const gridItems = imageGrid?.children;
      expect(gridItems).toHaveLength(6);
    });

    it('matches grid layout snapshot', () => {
      const { container } = renderWithRouter();

      // Optional snapshot test for grid layout
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Book Data Structure', () => {
    it('each book has required properties', () => {
      renderWithRouter();

      // Verify that each book link contains both image and title
      const links = screen.getAllByRole('link');

      links.forEach((link) => {
        // Each link should contain an image
        const image = link.querySelector('img');
        expect(image).toBeInTheDocument();

        // Each link should contain a title div
        const titleDiv = link.querySelector('div');
        expect(titleDiv).toBeInTheDocument();
        expect(titleDiv?.textContent).toBeTruthy();
      });
    });
  });
});
