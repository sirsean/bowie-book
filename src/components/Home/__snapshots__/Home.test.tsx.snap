// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home Component > Grid Layout > matches grid layout snapshot 1`] = `
<div
  class="_container_d6eec2"
>
  <h1
    class="_title_d6eec2"
  >
    Bowie's Books!
  </h1>
  <div
    class="_imageGrid_d6eec2"
  >
    <div
      class="_gridItem_d6eec2"
    >
      <a
        href="/bonne-adventure"
      >
        <img
          alt="Bonne Adventure Cover"
          loading="lazy"
          src="/books/bonne-adventure/0-cover.webp"
        />
        <div
          class="_gridItemTitle_d6eec2"
        >
          Bonne Adventure
        </div>
      </a>
    </div>
    <div
      class="_gridItem_d6eec2"
    >
      <a
        href="/dragon-fighter"
      >
        <img
          alt="Dragon Fighter Cover"
          loading="lazy"
          src="/books/dragon-fighter/0-cover.webp"
        />
        <div
          class="_gridItemTitle_d6eec2"
        >
          Dragon Fighter
        </div>
      </a>
    </div>
    <div
      class="_gridItem_d6eec2"
    >
      <a
        href="/skyward-bound"
      >
        <img
          alt="Skyward Bound Cover"
          loading="lazy"
          src="/books/skyward-bound/0.webp"
        />
        <div
          class="_gridItemTitle_d6eec2"
        >
          Skyward Bound
        </div>
      </a>
    </div>
    <div
      class="_gridItem_d6eec2"
    >
      <a
        href="/ziggy-the-bunny"
      >
        <img
          alt="Ziggy the Bunny Cover"
          loading="lazy"
          src="/books/ziggy-the-bunny/0-cover.webp"
        />
        <div
          class="_gridItemTitle_d6eec2"
        >
          Ziggy the Bunny
        </div>
      </a>
    </div>
    <div
      class="_gridItem_d6eec2"
    >
      <a
        href="/super-bowie"
      >
        <img
          alt="Super Bowie Cover"
          loading="lazy"
          src="/books/super-bowie/0-cover.jpg"
        />
        <div
          class="_gridItemTitle_d6eec2"
        >
          Super Bowie
        </div>
      </a>
    </div>
    <div
      class="_gridItem_d6eec2"
    >
      <a
        href="/superkitty-saves-bunnytown"
      >
        <img
          alt="Superkitty Saves Bunnytown Cover"
          loading="lazy"
          src="/books/superkitty-saves-bunnytown/superkitty-cover.png"
        />
        <div
          class="_gridItemTitle_d6eec2"
        >
          Superkitty Saves Bunnytown
        </div>
      </a>
    </div>
  </div>
</div>
`;
